### PR TITLE
Add selector-not-notation

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

#5974

> Is there anything in the PR that needs further explanation?

read https://github.com/stylelint/stylelint/issues/5974#issuecomment-1071396510

### CODE REVIEW

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/__tests__/index.js:58]: e.g.
`:not(p)` discards _all_ p tags hence `:not(p):not(.foo)` would not match any p
`:not(p.foo)` may (if there are p left without a foo class)

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/index.js:29]: If another rule ever needs it, it will have to moved to utils.
YAGNI

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/index.js:96]: https://github.com/postcss/postcss-selector-not

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:2]: Let's add a description, example and extended description

https://stylelint.io/user-guide/rules/about#readmes

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:13]:
```suggestion
:not(a, div) {}
```

Let's align with [our conventions](https://stylelint.io/developer-guide/rules/#write-tests). There are lots of rules in Stylelint, so we try to keep things consistent to make things easier our users and for when we're maintaining.

I'll update the docs to reflect that we usually use `div` as the secondary type selector.

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:18]:
```suggestion
:not(a.foo) {}
```

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:25]:
```suggestion
:not(a):not(div) {}
```

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:30]:
```suggestion
:not(a) {}
```

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:39]:
```suggestion
:not(a):not(div) {}
```

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:51]:
```suggestion
:not(a.foo) {}
```

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [inline lib/rules/selector-not-notation/README.md:56]:
```suggestion
:not(a).foo:not(:empty) {}
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:15]:
```suggestion
			code: ':not( a ) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:18]:
```suggestion
			code: ':nOt(a) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:51]:
```suggestion
			code: ':not(a, div) {}',
			fixed: ':not(a):not(div) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:57]:
```suggestion
			code: ':not(a.foo) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:73]:
```suggestion
			code: ':not()::after {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:76]:
```suggestion
			code: ':not(a, div) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:79]:
```suggestion
			code: ':not(a.foo) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:82]:
```suggestion
			code: ':not(a).foo:not(:empty) {}',
```

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/__tests__/index.js:89]:
```suggestion
			code: ':not(.foo):not(a) {}',
			fixed: ':not(.foo, a) {} {}',
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@Mouvedia Thank you! This is looking great already.

I've suggested some minor nits in the tests and docs, but the rule itself is shaping up really nicely.

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/__tests__/index.js:73]: damn
one typo
good catch

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [reply lib/rules/selector-not-notation/README.md:2]: I'm going offline for the weekend now, but I'll try to rustle up an extended description early next week unless someone beats me to it.

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/__tests__/index.js:43]: typo

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/index.js:9]: will have to be fixed upstream
no rule except this one uses this util, so we are safe
the extension at line 30 is future proof

File: lib/rules/selector-not-notation/__tests__/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/__tests__/index.js:114]: not worth pursuing

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [inline lib/rules/selector-not-notation/index.js:174]: typescript should complain on this line, but doesn't

File: docs/user-guide/rules/list.md
Comment (jeddy3) [inline docs/user-guide/rules/list.md:201]:
```suggestion
- [`selector-not-notation`](../../../lib/rules/selector-not-notation/README.md): Specify simple or complex notation for `:not()` pseudo-classes (Autofixable).
```

To be consistent with other rule descriptions.

File: lib/rules/selector-not-notation/README.md
Comment (jeddy3) [reply lib/rules/selector-not-notation/README.md:2]: Let's try (taken from [this spec](https://www.w3.org/TR/selectors-4/#negation)):

---

# selector-not-notation

Specify simple or complex notation for `:not()` pseudo-classes.

```css
    a:not(.foo, .bar) {}
/**  ↑
 * This notation */
```

In Selectors Level 3, only a single _simple selector_ was allowed as the argument to `:not()`, whereas Selectors Level 4 allows a _selector list_.

Use:

- `"complex"` to author modern Selectors Level 4 CSS
-  `"simple"` for backwards compatibility with older browsers

The [`fix` option](../../../docs/user-guide/usage/options.md#fix) can automatically fix most of the problems reported by this rule.

---

File: lib/rules/selector-not-notation/index.js
Comment (jeddy3) [inline lib/rules/selector-not-notation/index.js:90]: Let's return early here if there's no `:not` so we avoid calling the selector parser:

```js
if (!selector.includes(':not')) return;
```

And let's check that the [selector is standard](https://github.com/stylelint/stylelint/blob/main/lib/utils/isStandardSyntaxSelector.js) too to avoid issues with interpolation and the like.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@Mouvedia Thanks! It's looking great, especially the autofix.

I've requested some minor changes.

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/index.js:90]: Hey, sorry for the delay. Ill make the requested changes in the upcoming days.

I just saw PR 5983; is it related to my [comment](https://github.com/stylelint/stylelint/pull/5975#discussion_r831598929)?

**Review by jeddy3 — APPROVED** [review]
Thanks for making the changes.

LGTM, thank you!

(Pull requests that add rules need a second review, so we'll wait on one of those before merging, as the second pair of eyes usually catches something the first pair missed).

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:52]:
```suggestion
 * @returns {node is Pseudo}
```

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:102]:
```suggestion
						if (context.fix) return fixComplex(prev);
```

**Review by ybiquitous — COMMENTED** [review]
[suggest] What if making `isNot()` an assert function? Type-casting via `@type` will be unnecessary.

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:8]:
```suggestion
const validateOptions = require('../../utils/validateOptions');
const { assert } = require('../../utils/validateTypes');
```

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:148]:
```suggestion
	const firstSelector = simpleSelectors.shift();

	assert(firstSelector);

	not.nodes.push(firstSelector);
```

**Review by ybiquitous — COMMENTED** [review]
[suggest] `.shift()` can return `undefined`, so TS compiler complains. What if checking `.shift()` return value by `assert()` and removing `@ts-ignore`?

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:153]: [suggest] We can remove type-cast via `assert()` like this:

```suggestion
	assert(not.parent);

	for (const s of simpleSelectors) {
		const last = not.parent.last;

		not.parent.insertAfter(
```

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [inline lib/rules/selector-not-notation/index.js:182]: [suggest] If `isNot()` is an assertion function, we can remove `@type` also here:

```suggestion
		const selectors = node.nodes;
```

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [reply lib/rules/selector-not-notation/index.js:178]: If `noUncheckedIndexedAccess` is enabled, TS compiler will do it. See #5983.

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [reply lib/rules/selector-not-notation/index.js:10]: I've opened PR https://github.com/postcss/postcss-selector-parser/pull/256 to fix the bug.

Comment (ybiquitous) [reply lib/rules/selector-not-notation/index.js:10]: [suggest] What if making the comment TODO like this?

```suggestion
	isPseudoClass: buggyIsPseudoClass, // TODO: https://github.com/postcss/postcss-selector-parser/pull/256
```

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/index.js:52]: This function is not checking whether this is a pseudo or not.
Check its four references.

Is it unclear? Should it be renamed to `isNotPseudoClass` or add a description in the jsdoc comment?

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/index.js:102]: After line 98 we known we have a `Pseudo` but `prev()` returns a `Node` hence the casting so that it doesn't complain.

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/index.js:178]: In the meantime I have added `if (!head.nodes.length) return;`.

File: lib/rules/selector-not-notation/index.js
Comment (Mouvedia) [reply lib/rules/selector-not-notation/index.js:180]: Can you give me an example of what you mean by "assertion function"?

Comment (ybiquitous) [reply lib/rules/selector-not-notation/index.js:180]: Ah, sorry. I mean a "type guard function", not "assertion function". 🙇🏼

File: lib/rules/selector-not-notation/index.js
Comment (ybiquitous) [reply lib/rules/selector-not-notation/index.js:52]: `postcss-selector-parser`'s original `isPseudoClass()` type-guard function checks that a specified node is a `Pseudo` type or not. The below is the definition:

```ts
function isPseudoClass(node: any): node is Pseudo;
```

So, I think we can make our `isNot()` (and also `isPseudoClass()`) a type-guard function.
The naming of `isNot()` is not my concern, but feel free to rename it.

File: lib/rules/selector-not-notation/README.md
Comment (Mouvedia) [reply lib/rules/selector-not-notation/README.md:2]: > the argument _to_

of?

**Review by ybiquitous — APPROVED** [review]
@Mouvedia Thank you. LGTM 👍🏼

File: lib/rules/selector-not-notation/README.md
Comment (Mouvedia) [reply lib/rules/selector-not-notation/README.md:2]: > the argument to

@jeddy3 of?

Comment (jeddy3) [reply lib/rules/selector-not-notation/README.md:2]: Let's keep it as it is as it's a copy-paste job from the spec.

**Comment by jeddy3** [comment]
Changelog entry added:

> - Added: `selector-not-notation` rule ([#5975](https://github.com/stylelint/stylelint/pull/5975)).

@Mouvedia Thanks again for contributing this rule.

**Comment by jeddy3** [comment]
@ybiquitous Thanks for the 2nd review. I'm glad you reviewed as your type checking suggestions are always fantastic!

**Pull request merged**
